ukvm/ukvm-core.c

```diff
@@ -224,45 +224,55 @@ static void load_code(const char *file, uint8_t *mem,     /* IN */
      * ALIGN_UP(p_memsz, p_align) bytes on memory.
      */
     for (ph_i = 0; ph_i < ph_cnt; ph_i++) {
-        uint8_t *dst;
-        size_t _end;
+        uint8_t *daddr;
+        uint64_t _end;
         size_t offset = phdr[ph_i].p_offset;
         size_t filesz = phdr[ph_i].p_filesz;
         size_t memsz = phdr[ph_i].p_memsz;
         uint64_t paddr = phdr[ph_i].p_paddr;
         uint64_t align = phdr[ph_i].p_align;
+        uint64_t result;
 
-        if ((phdr[ph_i].p_type & PT_LOAD) == 0)
+        if (phdr[ph_i].p_type != PT_LOAD)
             continue;
 
         /* XXX this won't work after having the two memory slots */
         assert(GUEST_SIZE < KVM_32BIT_GAP_SIZE);
-        dst = mem + paddr;
-
-        numb = pread_in_full(fd_kernel, dst, filesz, offset);
-        if (numb < 0)
-            err(1, "%s: Read error", file);
-        if ((size_t) numb != filesz)
-            errx(1, "%s: Invalid segment", file);
-
-        memset(mem + paddr + filesz, 0, memsz - filesz);
-
-        /* Protect the executable code */
-        if (phdr[ph_i].p_flags & PF_X)
-            mprotect((void *) dst, memsz, PROT_EXEC | PROT_READ);
 
+        if ((paddr >= GUEST_SIZE) || add_overflow(paddr, filesz, result) ||
+            (result >= GUEST_SIZE)) {
+            errx(1, "%s: Invalid segment: paddr=0x%" PRIx64 ", filesz=%zu",
+                    file, paddr, filesz);
+        }
+        if (add_overflow(paddr, memsz, result) || (result >= GUEST_SIZE)) {
```

Comment:
I'm not 100% sure this can't overflow after the ALIGN_UP below. It doesn't seem to because (but wanted to double check with you): if `(paddr_memsz <= 1e64)` then `ALIGN_UP(paddr+memsz, 4096) <= 1e64`. And that will be true if paddr and memsz are both `uint64_t` so `add_overflow` becomes `UaddLL_overflow`, and if the check is `<=` and not `<`. Wouldn't it be safer to make memsz an uint64_t just to be sure the correct add_overflow is used?

`(result >= GUEST_SIZE)` is not needed here as that check is done one line below.